jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java

```diff
@@ -489,72 +489,93 @@ public Collection<Part> getParts() throws IOException, ServletException
+                long maxFormContentSize = servletContextHandler.getMaxFormContentSize();
+                int maxFormKeys = servletContextHandler.getMaxFormKeys();
 
-                    // Servlet Spec 3.0 pg 23, parts without filename must be put into params.
-                    String charset = null;
-                    if (p.getContentType() != null)
-                        charset = MimeTypes.getCharsetFromContentType(p.getContentType());
+                long formContentSize = 0;
+                int count = 0;
+                for (Part p : parts)
+                {
+                    if (maxFormKeys > 0 && ++count > maxFormKeys)
+                        throw new IllegalStateException("Too many form keys > " + maxFormKeys);
 
-                    try (InputStream is = p.getInputStream())
+                    if (p.getSubmittedFileName() == null)
                     {
-                        String content = IO.toString(is, charset == null ? defaultCharset : Charset.forName(charset));
-                        if (_contentParameters == null)
-                            _contentParameters = new Fields();
-                        _contentParameters.add(p.getName(), content);
+                        formContentSize = Math.addExact(formContentSize, p.getSize());
+                        if (maxFormContentSize >= 0 && formContentSize > maxFormContentSize)
+                            throw new IllegalStateException("Form is larger than max length " + maxFormContentSize);
+
+                        // Servlet Spec 3.0 pg 23, parts without filename must be put into params.
+                        String charset = null;
+                        if (p.getContentType() != null)
+                            charset = MimeTypes.getCharsetFromContentType(p.getContentType());
+
+                        try (InputStream is = p.getInputStream())
+                        {
+                            String content = IO.toString(is, charset == null ? defaultCharset : Charset.forName(charset));
+                            if (_contentParameters == null)
+                                _contentParameters = new Fields();
+                            _contentParameters.add(p.getName(), content);
+                        }
                     }
                 }
             }
+            catch (Throwable t)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("getParts", t);
+
+                Throwable cause;
+                if (t instanceof ExecutionException ee)
+                    cause = ee.getCause();
+                else if (t instanceof ServletException se)
+                    cause = se.getCause();
+                else
+                    cause = t;
+
+                if (cause instanceof IOException ioException)
+                    throw ioException;
+
+                throw new ServletException(new BadMessageException("bad multipart", cause));
```

Comment:
Because Servlet API says that is what we should throw to allow app to catch and handle.   We wrap a BadMessage to make sure we send a 400 if the app does not catch.
Yes it is cumbersome... and probably fragile with regards to the TCK for failures.